### PR TITLE
feat(sdk-client): validate arguments

### LINKED ISSUE #41: [sdk-client] validate request shape and function arguments
- `exec` (request shape)
- `process` (request shape, must be a GET request, fn must be a function)

### PR DESCRIPTION
**Pull request by wizzy25**
affects: @commercetools/sdk-client

validate the request shape and function arguments in the `exec` and `process` methods

resolves #41

### CODE REVIEW

File: packages/sdk-client/src/client.js [hunk 1]
```diff
@@ -135,3 +142,16 @@ function compose (...funcs: Array<Function>): Function {
 
   return funcs.reduce((a, b) => (...args) => a(b(...args)))
 }
+
+function validate (request: Object, fn?: ProcessFn) {
+  if (!request)
+    throw new Error('Missing required `request` argument.')
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: Let's try to provide meaningful and documented error messages 😉

For example:

> The `exec` function requires a `Request` object as an argument. See https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest

File: packages/sdk-client/src/client.js [hunk 2]
```diff
@@ -135,3 +142,16 @@ function compose (...funcs: Array<Function>): Function {
...
+  if (typeof request.uri !== 'string' || !request.method)
+    throw new Error('Request is invalid.')
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: > The `Request` object provided to the `exec` function is missing required fields. See https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest

File: packages/sdk-client/src/client.js [hunk 3]
```diff
@@ -135,3 +142,16 @@ function compose (...funcs: Array<Function>): Function {
 
   return funcs.reduce((a, b) => (...args) => a(b(...args)))
 }
+
+function validate (request: Object, fn?: ProcessFn) {
+  if (!request)
+    throw new Error('Missing required `request` argument.')
+  if (typeof request.uri !== 'string' || !request.method)
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: Nitpick: we could also validate that the `method` is one of the available `http` methods.

See `require('http').METHODS`

File: packages/sdk-client/src/client.js [hunk 4]
```diff
@@ -135,3 +142,16 @@ function compose (...funcs: Array<Function>): Function {
...
+    throw new Error('Request is invalid.')
+  if (fn) {
+    if (request.method !== 'GET')
+      throw new Error('Invalid request type. Must be a `GET` request')
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: > The `process` function only works for `Request` of type `GET`. See https://commercetools.github.io/nodejs/sdk/api/sdkClient.html#processrequest-processfn-options

File: packages/sdk-client/src/client.js [hunk 5]
```diff
@@ -135,3 +142,16 @@ function compose (...funcs: Array<Function>): Function {
...
+    if (typeof fn !== 'function')
+      throw new Error('Invalid argument! Second argument must be a function')
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: > The `process` function accepts a `Function` as a second argument that returns a `Promise`. See https://commercetools.github.io/nodejs/sdk/api/sdkClient.html#processrequest-processfn-options

File: packages/sdk-client/src/client.js [hunk 6]
```diff
@@ -71,6 +73,11 @@ export default function createClient (options: ClientOptions): Client {
       fn: ProcessFn,
       opt: ProcessOptions = { accumulate: true },
     ): Promise<Array<Object>> {
+      if (!fn)
+        throw new Error('Missing argument! Second argument must be defined')
```
Comment (emmenko) [inline packages/sdk-client/src/client.js]: Do we still need this check?

Comment (wizzy25) [inline packages/sdk-client/src/client.js]: if the `function` is required, it's presence cannot be validated in the `validate` function. How would you suggest I do this?

File: packages/sdk-client/src/client.js (hunk 3 above)
Comment (emmenko) [inline packages/sdk-client/src/client.js]: Why not, this client potentially works with any kind of requests ;)

File: packages/sdk-client/src/client.js (hunk 6 above)
Comment (emmenko) [inline packages/sdk-client/src/client.js]: 🤔 How about validating only the request in a separate function, then validate `fn` directly here.

```js
// for "exec" method
validateRequest(request)

// for "process" method
validateRequest(request, { allowedMethods: ['GET'] })
if (!fn)
  throw new Error('The process function accepts a Function as a second argument that returns a Promise. See https://commercetools.github.io/nodejs/sdk/api/sdkClient.html#processrequest-processfn-options')

// function signature
function validateRequest (request, options) {
}
```

**Comment by wizzy25** [comment]
@emmenko made some changes... also moved the validate function to another file :-)

File: packages/sdk-client/src/validate.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  ) {
+  if (!request)
+    // eslint-disable-next-line max-len
+    throw new Error(`The "${funcName}" function requires a "Request" object as an argument. See https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest`)
+
+  if (typeof request.uri !== 'string')
+    // eslint-disable-next-line max-len
+    throw new Error(`The "${funcName}" Request object requires a valid uri. See https://commercetools.github.io/nodejs/sdk/Glossary.html#clientrequest`)
+
+  if (options.allowedMethods.indexOf(request.method) === -1)
```
Comment (emmenko) [inline packages/sdk-client/src/validate.js]: Nitpick: you can use `options.allowedMethods.includes(request.method)` 😉

**Pull request merged**